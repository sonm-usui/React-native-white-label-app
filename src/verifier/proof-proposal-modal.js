// @flow
import React, { useCallback, useMemo } from 'react'
import { useDispatch, useSelector } from 'react-redux'
import { View, StyleSheet, StatusBar, ScrollView, FlatList } from 'react-native'
import { verticalScale, moderateScale } from 'react-native-size-matters'
import { homeDrawerRoute, homeRoute, proofProposalRoute } from '../common'
import type { ReactNavigation } from '../common/type-common'
import { colors, fontSizes, fontFamily } from '../common/styles'
import { ExpandableText } from '../components/expandable-text/expandable-text'
import { modalOptions } from '../connection-details/utils/modalOptions'
import { getLockStore, getVerifier } from '../store/store-selector'
import { Loader } from '../components'
import type { InvitationPayload } from '../invitation/type-invitation'
import type { AriesPresentationPreviewAttribute } from '../proof-request/type-proof-request'
import { ModalButtons } from '../components/buttons/modal-buttons'
import { ModalHeader } from '../connection-details/components/modal-header'
import { CONNECTION_INVITE_TYPES } from '../invitation/type-invitation'
import { acceptOutOfBandInvitation } from '../invitation/invitation-store'
import { proofProposalAccepted } from './verifier-store'
import { authForAction } from '../lock/lock-auth-for-action'
import {
  CustomProofProposalModal,
  proofProposalHeadline,
  proofProposalAcceptButtonText,
  proofProposalDenyButtonText,
} from '../external-imports'
import { unlockApp } from '../lock/lock-store'

export type ProofProposalProps = {
  backRedirectRoute?: string | null,
  uid: string,
  invitationPayload: InvitationPayload,
  senderName: string,
}

export const ProofProposalComponent = ({
  navigation,
  route: { params },
}: ReactNavigation) => {
  const {
    uid,
    backRedirectRoute,
    senderName,
    invitationPayload,
  }: ProofProposalProps = params

  const dispatch = useDispatch()

  const verifier = useSelector((state) => getVerifier(state, uid))
  const lock = useSelector(getLockStore)

  const presentationProposal = useMemo(() => verifier.presentationProposal, [
    verifier,
  ])

  const onAcceptAuthSuccess = useCallback(() => {
    if (invitationPayload.type === CONNECTION_INVITE_TYPES.ARIES_OUT_OF_BAND) {
      dispatch(acceptOutOfBandInvitation(invitationPayload))
    } else {
      dispatch(proofProposalAccepted(uid))
    }
    navigation.navigate(homeRoute, {
      screen: homeDrawerRoute,
      params: undefined,
    })
  }, [uid, invitationPayload])

  const onAccept = useCallback(() => {
    authForAction({
      lock,
      navigation,
      onSuccess: onAcceptAuthSuccess,
      unlockApp: () => dispatch(unlockApp()),
    })
  }, [onAcceptAuthSuccess])

  const onDeny = useCallback(() => {
    if (backRedirectRoute) {
      navigation.navigate(backRedirectRoute)
    } else {
      navigation.goBack(null)
    }
  }, [backRedirectRoute])

  const keyExtractor = (_: any, index: number) => index.toString()

  const renderItem = (attribute: AriesPresentationPreviewAttribute) => (
    <View style={styles.attributeWrapper}>
      <ExpandableText text={attribute.name} style={styles.attribute} />
    </View>
  )

  const acceptBtnText = proofProposalAcceptButtonText || 'Accept'
  const denyButtonText = proofProposalDenyButtonText || 'Cancel'

  return !presentationProposal ? (
    <Loader />
  ) : (
    <>
      <View style={styles.contentWrapper}>
        <ScrollView showsVerticalScrollIndicator={false}>
          <StatusBar
            backgroundColor={colors.black}
            barStyle={'light-content'}
          />
          <ModalHeader
            institutionalName={senderName}
            credentialName={presentationProposal.comment || 'Proof Proposal'}
            credentialText={'Offers you to present data'}
            imageUrl={verifier.senderLogoUrl}
          />
          <FlatList
            data={presentationProposal.presentation_proposal.attributes}
            keyExtractor={keyExtractor}
            renderItem={({ item }) => renderItem(item)}
          />
        </ScrollView>
      </View>
      <ModalButtons
        onPress={onAccept}
        onIgnore={onDeny}
        acceptBtnText={acceptBtnText}
        denyButtonText={denyButtonText}
        colorBackground={colors.main}
        secondColorBackground={colors.red}
      />
    </>
  )
}

const screen =
  (CustomProofProposalModal && CustomProofProposalModal.screen) ||
  ProofProposalComponent

const headline = proofProposalHeadline || 'Proof Proposal'

const navigationOptions =
  (CustomProofProposalModal && CustomProofProposalModal.navigationOptions) ||
  modalOptions(headline, 'CloseIcon')

export const ProofProposalModal = {
  routeName: proofProposalRoute,
  screen,
}

ProofProposalModal.screen.navigationOptions = navigationOptions

const styles = StyleSheet.create({
  contentWrapper: {
    flex: 1,
    paddingLeft: '5%',
    paddingRight: '5%',
  },
  attributeWrapper: {
    paddingTop: moderateScale(12),
    borderBottomColor: colors.gray4,
    borderBottomWidth: StyleSheet.hairlineWidth,
    paddingBottom: moderateScale(10),
  },
  attribute: {
    fontSize: verticalScale(fontSizes.size5),
    color: colors.black,
    width: '100%',
    textAlign: 'left',
    fontFamily: fontFamily,
    fontWeight: '700',
    marginBottom: moderateScale(2),
  },
})
