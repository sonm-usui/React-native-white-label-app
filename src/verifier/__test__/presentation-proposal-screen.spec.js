// @flow

import React from 'react'
import { Provider } from 'react-redux'
import 'react-native'
import renderer from 'react-test-renderer'

import {
  getNavigation,
  defaultUUID,
  getStore,
} from '../../../__mocks__/static-data'
import merge from 'lodash.merge'
import { ProofProposalComponent } from '../proof-proposal-modal'

const getProps = () => {
  return {
    navigation: {
      ...getNavigation(),
    },
    route: {
      params: {
        uid: defaultUUID,
        invitationPayload: {},
        attachedRequest: {},
        senderName: 'Test',
      },
    },
  }
}

const getState = () => {
  const currentState = getStore().getState()
  return {
    ...getStore(),
    getState() {
      return merge(
        {},
        {
          ...currentState,
          verifier: {
            [defaultUUID]: {
              presentationProposal: {
                "@type": "https://didcomm.org/present-proof/1.0/propose-presentation",
                "@id": "uuid",
                "comment": "some comment",
                "presentation_proposal": {
                  "@type": "https://didcomm.org/present-proof/1.0/presentation-preview",
                  "attributes": [
                    {
                      "name": "First Name",
                    },
                    {
                      "name": "Last Name",
                    },
                  ]
                }
              },
              senderLogoUrl: '',
            }
          }
        }
      )
    },
  }
}

const setup = (currentStore) => {
  const props = getProps()
  const component = renderer.create(
    <Provider store={currentStore}>
      <ProofProposalComponent {...props}/>
    </Provider>
  )
  const instance = component.getInstance()

  return { props, component, instance }
}

describe('<PresentationProposalComponent />', () => {
  it('should match snapshot', () => {
    const { component } = setup(getState())
    expect(component.toJSON()).toMatchSnapshot()
  })
})
