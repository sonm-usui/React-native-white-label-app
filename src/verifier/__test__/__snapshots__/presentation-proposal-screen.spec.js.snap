// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<PresentationProposalComponent /> should match snapshot 1`] = `
Array [
  <View
    style={
      Object {
        "flex": 1,
        "paddingLeft": "5%",
        "paddingRight": "5%",
      }
    }
  >
    <RCTScrollView
      showsVerticalScrollIndicator={false}
    >
      <View>
        <View
          style={
            Object {
              "alignItems": "center",
              "backgroundColor": "#FFFFFF",
              "flexDirection": "column",
              "paddingRight": "5%",
            }
          }
        >
          <View
            style={
              Object {
                "flexDirection": "row",
              }
            }
          >
            <View
              style={
                Object {
                  "flex": 1,
                  "paddingBottom": 15.314285714285713,
                  "paddingTop": 14.357142857142858,
                }
              }
            >
              <Text
                style={
                  Object {
                    "color": "#777777",
                    "fontFamily": "System",
                    "fontSize": 11.117647058823529,
                    "fontWeight": "400",
                    "paddingBottom": 15.314285714285713,
                    "textAlign": "center",
                  }
                }
              >
                Offers you to present data
              </Text>
              <Text
                ellipsizeMode="tail"
                numberOfLines={2}
                onTextLayout={[Function]}
                style={
                  Object {
                    "color": "#505050",
                    "fontFamily": "System",
                    "fontSize": 11.911764705882351,
                    "fontWeight": "700",
                    "textAlign": "center",
                  }
                }
              >
                Test
              </Text>
            </View>
          </View>
          <View
            style={
              Object {
                "borderRadius": 45.94285714285714,
              }
            }
          >
            <View
              style={
                Object {
                  "alignItems": "center",
                  "backgroundColor": "#777777",
                  "borderRadius": 38.285714285714285,
                  "height": 76.57142857142857,
                  "justifyContent": "center",
                  "width": 76.57142857142857,
                }
              }
            >
              <Text
                accessibilityLabel="default-logo"
                accessible={true}
                style={
                  Object {
                    "color": "#FFFFFF",
                    "fontFamily": "System",
                    "fontSize": 48,
                    "fontWeight": "bold",
                  }
                }
                testID="default-logo"
              >
                T
              </Text>
            </View>
          </View>
          <View
            style={
              Object {
                "borderBottomColor": "#505050",
                "borderBottomWidth": 0.5,
                "justifyContent": "center",
                "marginBottom": 11.485714285714286,
                "paddingBottom": 11.485714285714286,
                "paddingTop": 22.97142857142857,
                "width": "100%",
              }
            }
          >
            <Text
              ellipsizeMode="tail"
              numberOfLines={2}
              onTextLayout={[Function]}
              style={
                Object {
                  "color": "#505050",
                  "fontFamily": "System",
                  "fontSize": 15.088235294117647,
                  "fontWeight": "400",
                  "textAlign": "center",
                }
              }
            >
              some comment
            </Text>
          </View>
        </View>
        <RCTScrollView
          data={
            Array [
              Object {
                "name": "First Name",
              },
              Object {
                "name": "Last Name",
              },
            ]
          }
          getItem={[Function]}
          getItemCount={[Function]}
          keyExtractor={[Function]}
          onContentSizeChange={[Function]}
          onLayout={[Function]}
          onMomentumScrollBegin={[Function]}
          onMomentumScrollEnd={[Function]}
          onScroll={[Function]}
          onScrollBeginDrag={[Function]}
          onScrollEndDrag={[Function]}
          removeClippedSubviews={false}
          renderItem={[Function]}
          scrollEventThrottle={50}
          stickyHeaderIndices={Array []}
          viewabilityConfigCallbackPairs={Array []}
        >
          <View>
            <View
              onLayout={[Function]}
              style={null}
            >
              <View
                style={
                  Object {
                    "borderBottomColor": "#EAEAEA",
                    "borderBottomWidth": 0.5,
                    "paddingBottom": 9.571428571428571,
                    "paddingTop": 11.485714285714286,
                  }
                }
              >
                <Text
                  ellipsizeMode="tail"
                  numberOfLines={2}
                  onTextLayout={[Function]}
                  style={
                    Object {
                      "color": "#000000",
                      "fontFamily": "System",
                      "fontSize": 11.911764705882351,
                      "fontWeight": "700",
                      "marginBottom": 1.9142857142857141,
                      "textAlign": "left",
                      "width": "100%",
                    }
                  }
                >
                  First Name
                </Text>
              </View>
            </View>
            <View
              onLayout={[Function]}
              style={null}
            >
              <View
                style={
                  Object {
                    "borderBottomColor": "#EAEAEA",
                    "borderBottomWidth": 0.5,
                    "paddingBottom": 9.571428571428571,
                    "paddingTop": 11.485714285714286,
                  }
                }
              >
                <Text
                  ellipsizeMode="tail"
                  numberOfLines={2}
                  onTextLayout={[Function]}
                  style={
                    Object {
                      "color": "#000000",
                      "fontFamily": "System",
                      "fontSize": 11.911764705882351,
                      "fontWeight": "700",
                      "marginBottom": 1.9142857142857141,
                      "textAlign": "left",
                      "width": "100%",
                    }
                  }
                >
                  Last Name
                </Text>
              </View>
            </View>
          </View>
        </RCTScrollView>
      </View>
    </RCTScrollView>
  </View>,
  <View
    style={
      Array [
        Object {
          "backgroundColor": "#FFFFFF",
          "flexDirection": "column",
          "maxWidth": "100%",
          "padding": 14.357142857142858,
          "paddingBottom": 9.571428571428571,
          "width": "100%",
        },
        undefined,
      ]
    }
  >
    <View
      accessible={true}
      collapsable={false}
      focusable={true}
      nativeID="animatedComponent"
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "alignItems": "center",
          "backgroundColor": "#86B93B",
          "borderRadius": 5,
          "flexDirection": "row",
          "justifyContent": "center",
          "marginBottom": 15.314285714285713,
          "minWidth": "35%",
          "opacity": 1,
          "overflow": "hidden",
          "padding": 16.271428571428572,
          "paddingLeft": 9.571428571428571,
          "paddingRight": 9.571428571428571,
          "width": "100%",
        }
      }
    >
      <Text
        allowFontScaling={false}
        style={
          Array [
            Object {
              "fontSize": 19,
            },
            Object {
              "fontFamily": "System",
            },
            Object {
              "backgroundColor": "#333333",
              "color": "#ffffff",
            },
            Object {
              "fontWeight": "800",
            },
            Object {
              "textAlign": "center",
            },
            Object {
              "backgroundColor": "transparent",
            },
            null,
            null,
            null,
            null,
            null,
            null,
            null,
            null,
            null,
            null,
          ]
        }
        suppressHighlighting={true}
      >
        Accept
      </Text>
    </View>
    <View
      accessible={true}
      collapsable={false}
      focusable={true}
      nativeID="animatedComponent"
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "opacity": 1,
        }
      }
    >
      <View
        style={
          Array [
            Object {
              "alignItems": "stretch",
              "borderRadius": 5,
              "flexDirection": "row",
              "justifyContent": "center",
              "overflow": "hidden",
              "width": "100%",
            },
            Object {
              "alignItems": "center",
              "borderColor": "#CE0B24",
              "borderWidth": 1,
              "justifyContent": "center",
              "minWidth": "35%",
              "padding": 16.271428571428572,
              "paddingLeft": 9.571428571428571,
              "paddingRight": 9.571428571428571,
            },
          ]
        }
      >
        <Text
          allowFontScaling={false}
          style={
            Array [
              Object {
                "fontSize": 19,
              },
              Object {
                "fontFamily": "System",
              },
              Object {
                "backgroundColor": "#333333",
                "color": "#ffffff",
              },
              Object {
                "fontWeight": "600",
              },
              null,
              Object {
                "backgroundColor": "transparent",
              },
              null,
              Object {
                "color": "#d0021b",
              },
              null,
              null,
              null,
              null,
              null,
              null,
              null,
              null,
            ]
          }
          suppressHighlighting={true}
        >
          Cancel
        </Text>
      </View>
    </View>
  </View>,
]
`;
